packages/indexer-common/src/graph-node.ts

```diff
@@ -625,6 +718,231 @@ export class GraphNode {
+            (status) => status.subgraphDeployment.ipfsHash === dependency.base.ipfsHash,
+          )
+          if (!deploymentStatus) {
+            this.logger.error(`Subgraph not found in indexing status`, {
+              subgraph: dependency.base.ipfsHash,
+              indexingStatus,
+            })
+            throw new Error(`Subgraph not found in indexing status`)
+          } else {
+            this.logger.info(
+              'Dependency subgraph found, will try to sync it to the block required',
+              {
+                deploymentStatus,
+              },
+            )
+          }
+        } else if (!dependencyAssignment) {
+          this.logger.debug(
+            'Dependency subgraph not found, creating, deploying and pausing...',
+            {
+              name,
+              deployment: dependency.base.display,
+              block_required: dependency.block,
+            },
+          )
+          // are we paused at the block we wanted?
+
+          await this.create(name)
+          await this.deploy(name, dependency.base)
+        }
+        await this.syncToBlockAndPause(
+          dependency.block,
+          dependency.base,
+          subgraphChainName,
+        )
+      }
+    }
+  }
+
+  /**
+   * Wait for the block to be synced, polling indexing status until it is
+   * The Deployment should already be created and deployed to graph-node
+   * This will resume a paused subgraph if the block height target is higher than the
+   * current block height
+   */
+  public async syncToBlockAndPause(
+    blockHeight: number,
+    subgraphDeployment: SubgraphDeploymentID,
+    chainName: string | null,
+  ): Promise<void> {
+    async function waitForMs(ms: number) {
+      return new Promise((resolve) => setTimeout(resolve, ms))
+    }
+
+    this.logger.info(`Begin syncing subgraph deployment to block`, {
+      subgraph: subgraphDeployment.ipfsHash,
+      blockHeight,
+    })
+
+    // loop-wait for the block to be synced
+    for (;;) {
+      // first ensure it's been deployed and is active, or already paused
+      let deployed: SubgraphDeploymentAssignment[] = []
+      let attempt = 0
+
+      const maxAttempts = 5
+      const waitTimeMs = 3000
+
+      // wait and poll for the assignment to be created.
+      while (attempt < maxAttempts) {
+        await waitForMs(waitTimeMs)
+        deployed = await this.subgraphDeploymentAssignmentsByDeploymentID(
+          SubgraphStatus.ALL,
+          [subgraphDeployment.ipfsHash],
+        )
+        if (deployed.length > 0) {
+          this.logger.info(`Subgraph deployment active or already paused`, {
+            subgraph: subgraphDeployment.ipfsHash,
+            status: deployed,
+          })
+          break
+        }
+        this.logger.info(`Subgraph deployment not yet active, waiting...`, {
+          subgraph: subgraphDeployment.ipfsHash,
+          attempt,
+          deployed,
+        })
+        attempt += 1
```

Comment:
Am I reading this correctly: the loop-wait will try 5 times, with a 3 second pause in between attempts, after which it will fail out? That doesn't leave much time for the dependency to sync does it?

Reply:
This does what you're saying but only until graph-node reports the deployment as active, not finished syncing. If after that time has passed and graph-node still isn't showing the deployment is active, then we end things.